core-plugins/src/main/java/co/cask/hydrator/plugin/batch/joiner/Joiner.java

```diff
@@ -0,0 +1,289 @@
+@Name("Joiner")
+@Description("Performs join operation on records from each input based on required inputs. If all the inputs are " +
+  "required inputs, inner join will be performed. Otherwise inner join will be performed on required inputs and non " +
+  "records from non-required inputs will only be present if they match join criteria.")
+public class Joiner extends BatchJoiner<StructuredRecord, StructuredRecord, StructuredRecord> {
+  private final JoinerConfig conf;
+  private Map<String, Schema> inputSchemas;
+  private Schema outputSchema;
+  private Map<String, List<String>> perStageJoinKeys;
+  private Table<String, String, String> perStageFieldsToRename;
+  private Iterable<String> requiredInputs;
+  private static final Gson GSON = new Gson();
+  private static final Type MAP_STRING_MAP = new TypeToken<Map<String, Map<String, String>>>() {
+  }.getType();
+
+  private static final Gson GSON1 = new GsonBuilder()
+    .registerTypeAdapter(StructuredRecord.class, new StructuredRecordCodec())
+    .create();
+
+  public Joiner(JoinerConfig conf) {
+    this.conf = conf;
+  }
+
+  @Override
+  public void configurePipeline(PipelineConfigurer pipelineConfigurer) {
+    StageConfigurer stageConfigurer = pipelineConfigurer.getStageConfigurer();
+    Map<String, Schema> inputSchemas = stageConfigurer.getInputSchemas();
+    init();
+    //validate the input schema and get the output schema for it
+    stageConfigurer.setOutputSchema(getOutputSchema(inputSchemas));
+  }
+
+  @Override
+  public void prepareRun(BatchJoinerContext context) throws Exception {
+    if (conf.getNumPartitions() != null) {
+      context.setNumPartitions(conf.getNumPartitions());
+    }
+  }
+
+  @Override
+  public void initialize(BatchJoinerRuntimeContext context) throws Exception {
+    init();
+    inputSchemas = context.getInputSchemas();
+    outputSchema = context.getOutputSchema();
+  }
+
+  @Override
+  public StructuredRecord joinOn(String stageName, StructuredRecord record) throws Exception {
+    List<Schema.Field> fields = new ArrayList<>();
+    Schema schema = record.getSchema();
+
+    List<String> joinKeys = perStageJoinKeys.get(stageName);
+    int i = 1;
+    for (String joinKey : joinKeys) {
+      Schema.Field joinField = Schema.Field.of(String.valueOf(i++), schema.getField(joinKey).getSchema());
+      fields.add(joinField);
+    }
+    Schema keySchema = Schema.recordOf("join.key", fields);
+    StructuredRecord.Builder keyRecordBuilder = StructuredRecord.builder(keySchema);
+    i = 1;
+    for (String joinKey : joinKeys) {
+      keyRecordBuilder.set(String.valueOf(i++), record.get(joinKey));
+    }
+
+    return keyRecordBuilder.build();
+  }
+
+  @Override
+  public JoinConfig getJoinConfig() {
+    return new JoinConfig(requiredInputs);
+  }
+
+  @Override
+  public StructuredRecord merge(StructuredRecord joinKey, Iterable<JoinElement<StructuredRecord>> joinRow) {
+    StructuredRecord.Builder outRecordBuilder;
```

Comment:
can combine lines 131 and 132.
why separate the declaration and initialization?